### PR TITLE
Fix compiling process on osx by using static Qt libraries

### PR DESCRIPTION
**Pull request by matteoceraico**
New cmake arguments added:
- DISABLE_TESTS, disables all test
- QT_STATIC_BUILD, defines that qt is compiled as a static lib

Provided ability to exclude QT features, you need to add the parameters
(in accordance with QT) as a cmake argument (e.g.:
-DQT_NO_GRAPHICSVIEW=ON)

### CODE REVIEW

File: Source/cmake/OptionsQt.cmake [hunk 1]
```diff
@@ -200,9 +223,15 @@ set(REQUIRED_QT_VERSION 5.2.0)
 
 find_package(Qt5 ${REQUIRED_QT_VERSION} REQUIRED COMPONENTS Core Gui Network Sql)
 
-# FIXME: Allow building w/o these components
-find_package(Qt5OpenGL ${REQUIRED_QT_VERSION})
-find_package(Qt5Test ${REQUIRED_QT_VERSION} REQUIRED)
+if (ENABLE_OPENGL)
```
Comment (annulen) [inline Source/cmake/OptionsQt.cmake]: Not needed, `Qt5OpenGL` is not `REQUIRED`. `ENABLE_OPENGL` have different meaning, it's about OpenGL inside WebKit, and Qt5OpenGL is just about running in `QGLWidget`

File: Source/WebKit/PlatformQt.cmake [hunk 2]
```diff
@@ -1,3 +1,11 @@
+get_cmake_property(_variableNames VARIABLES)
+foreach (_variableName ${_variableNames})
+    if ((_variableName MATCHES "^QT_NO_") AND ("${${_variableName}}" STREQUAL "ON"))
```
Comment (annulen) [inline Source/WebKit/PlatformQt.cmake]: What do you do with this code? Looks like a wrong thing to do anyway

Comment (annulen) [inline Source/WebKit/PlatformQt.cmake]: OK, I see, you want to pass -D options right from command line. Use `CMAKE_CXX_FLAGS` for this purpose

File: Source/JavaScriptCore/shell/CMakeLists.txt [hunk 3]
```diff
@@ -41,7 +41,7 @@ if (WIN32)
     add_dependencies(jsc jscLib)
 endif ()
 
-if (NOT WIN32)
+if (NOT WIN32 AND ENABLE_TEST_SUPPORT)
```
Comment (annulen) [inline Source/JavaScriptCore/shell/CMakeLists.txt]: I think it should check ENABLE_API_TESTS instead

File: Source/cmake/OptionsQt.cmake [hunk 4]
```diff
@@ -51,7 +51,11 @@ option(ENABLE_TEST_SUPPORT "Build tools for running layout tests and related lib
 # Public options shared with other WebKit ports. There must be strong reason
 # to support changing the value of the option.
 WEBKIT_OPTION_DEFAULT_PORT_VALUE(ENABLE_ALLINONE_BUILD PUBLIC ON)
-WEBKIT_OPTION_DEFAULT_PORT_VALUE(ENABLE_API_TESTS PUBLIC ON)
+if (ENABLE_TEST_SUPPORT)
```
Comment (annulen) [inline Source/cmake/OptionsQt.cmake]: This options are not related to each other, you should disable them on command line separately

File: Source/cmake/OptionsQt.cmake [hunk 5]
```diff
@@ -115,6 +119,10 @@ set(ENABLE_WEBKIT ON)
 set(ENABLE_WEBKIT2 OFF)
 set(WTF_USE_UDIS86 1)
 
+if (QT_STATIC_BUILD)
```
Comment (annulen) [inline Source/cmake/OptionsQt.cmake]: 1. Don't use QT_STATIC_BUILD before get_target_property() is called, it's not intuitive

2. Raise error instead

```
if (QT_STATIC_BUILD AND SHARED_CORE)
  message(FATAL_ERROR "QT_STATIC_BUILD and SHARED_CORE options are incompatible")
endif ()

Comment (matteoceraico) [inline Source/cmake/OptionsQt.cmake]: get_target_property() will be moved before, i've done a mistake

File: Source/WebKit/PlatformQt.cmake [hunk 6]
```diff
@@ -213,6 +213,27 @@ list(APPEND WebKit_LIBRARIES
...
+                ${SECURITY_LIBRARY}
+                ${CARBON_LIBRARY}
+                ${COCOA_LIBRARY}
+        )
+    endif()
+
+    list(APPEND WebKit_LIBRARIES
+        PUBLIC
+            ${OPENSSL_LIBRARIES}
+            ${_qt5Widgets_install_prefix}/lib/libqtharfbuzzng.a
```
Comment (Vitallium) [inline Source/WebKit/PlatformQt.cmake]: Static libraries have extension `lib` on Windows. This will not work on Windows.

File: Source/WebKit/qt/WidgetSupport/QtFallbackWebPopup.cpp [hunk 7]
```diff
@@ -58,10 +58,14 @@ void QtFallbackWebPopup::show(const QWebSelectData& data)
     populate(data);
 
     QRect rect = geometry();
+#ifdef QT_NO_GRAPHICSVIEW
```
Comment (Vitallium) [inline Source/WebKit/qt/WidgetSupport/QtFallbackWebPopup.cpp]: We already have this fix in `qtwebkit-stable`. You should do a rebase.

**Comment by annulen** [comment]
There is Qt patch which makes Qt cmake files to link Qt's dependencies automatically: https://codereview.qt-project.org/#/c/33193, however it was reverted in upstream. You may want to implement similar logic on our side, so that we extract dependencies from `.prl` file instead of finding them

File: Source/WebCore/PlatformQt.cmake [hunk 8]
```diff
@@ -147,6 +147,13 @@ list(APPEND WebCore_SOURCES
     platform/text/qt/TextBreakIteratorInternalICUQt.cpp
 )
 
+if ((DEFINED ENV{SQLITE3SRCDIR}) AND (EXISTS $ENV{SQLITE3SRCDIR}/sqlite3.c))
+    list(APPEND WebCore_SOURCES
+        "$ENV{SQLITE3SRCDIR}/sqlite3.c"
+    )
+    add_definitions(-DSQLITE_CORE -DSQLITE_OMIT_LOAD_EXTENSION -DSQLITE_OMIT_COMPLETE)
```
Comment (annulen) [inline Source/WebCore/PlatformQt.cmake]: Looks like you've lost include directory now

Comment (matteoceraico) [inline Source/WebCore/PlatformQt.cmake]: I'm settting it on OptionsQt.cmake:
`if ((NOT DEFINED ENV{SQLITE3SRCDIR}) OR (NOT EXISTS $ENV{SQLITE3SRCDIR}/sqlite3.c))
    find_package(Sqlite REQUIRED)
else ()
    set(SQLITE_INCLUDE_DIR $ENV{SQLITE3SRCDIR})
endif ()`

Comment (annulen) [inline Source/WebCore/PlatformQt.cmake]: Whoa. In this case, please avoid duplication of this condition

File: Source/cmake/OptionsQt.cmake [hunk 9]
```diff
@@ -163,10 +172,19 @@ endif ()
 find_package(LibXml2 2.8.0 REQUIRED)
 find_package(JPEG REQUIRED)
 find_package(PNG REQUIRED)
-find_package(Sqlite REQUIRED)
+if ((NOT DEFINED ENV{SQLITE3SRCDIR}) OR (NOT EXISTS $ENV{SQLITE3SRCDIR}/sqlite3.c))
+    find_package(Sqlite REQUIRED)
+else ()
+    set(SQLITE_INCLUDE_DIR $ENV{SQLITE3SRCDIR})
+endif ()
```
Comment (annulen) [inline Source/cmake/OptionsQt.cmake]:
```
if (DEFINED ENV{SQLITE3SRCDIR})
     set(SQLITE_SOURCE_FILE $ENV{SQLITE3SRCDIR}/sqlite3.c)
     if (NOT EXISTS ${SQLITE_SOURCE_FILE}) -> FATAL_ERROR
     set(SQLITE_INCLUDE_DIR $ENV{SQLITE3SRCDIR})
endif
```

Comment (annulen) [inline Source/cmake/OptionsQt.cmake]: And add `SQLITE_SOURCE_FILE` to WebCore sources unconditionally

File: Source/JavaScriptCore/PlatformQt.cmake [hunk 10]
```diff
@@ -18,6 +18,12 @@ list(APPEND JavaScriptCore_LIBRARIES
     ${Qt5Core_LIBRARIES}
 )
 
+if (QT_STATIC_BUILD)
+    list(APPEND JavaScriptCore_LIBRARIES
+        ${ZLIB_LIBRARIES}
```
Comment (annulen) [inline Source/JavaScriptCore/PlatformQt.cmake]: Should not be needed anymore

File: Source/WebKit/PlatformQt.cmake [hunk 11]
```diff
@@ -213,6 +213,13 @@ list(APPEND WebKit_LIBRARIES
         ${Qt5Network_LIBRARIES}
 )
 
+if (QT_STATIC_BUILD)
+    list(APPEND WebKit_LIBRARIES
+        PUBLIC
+            ${ZLIB_LIBRARIES}
```
Comment (annulen) [inline Source/WebKit/PlatformQt.cmake]: Should not be needed anymore

File: Source/cmake/OptionsQt.cmake [hunk 12]
```diff
@@ -202,9 +241,26 @@ find_package(Qt5 ${REQUIRED_QT_VERSION} REQUIRED COMPONENTS Core Gui Network Sql
...
 # FIXME: Allow building w/o these components
 find_package(Qt5OpenGL ${REQUIRED_QT_VERSION})
-find_package(Qt5Test ${REQUIRED_QT_VERSION} REQUIRED)
 find_package(Qt5Widgets ${REQUIRED_QT_VERSION} REQUIRED)
 
+if (ENABLE_TEST_SUPPORT)
+    find_package(Qt5Test ${REQUIRED_QT_VERSION} REQUIRED)
+endif ()
+
+if (QT_STATIC_BUILD)
```
Comment (annulen) [inline Source/cmake/OptionsQt.cmake]: I think we should change Qt handling to be more like what is done for `glib` and `gstreamer` in this very file. I.e. add variable `QT_COMPONENTS`, add modules to it as we go, and after all conditions do 2 things:
```
find_package(Qt5 ${REQUIRED_QT_VERSION} REQUIRED COMPONENTS ${QT_COMPONENTS})
foreach (qt_module ${QT_COMPONENTS})
    macro_process_qtbase_prl_file(Qt5::${qt_module})
endforeach ()
```
Please check if syntax is right

File: Source/JavaScriptCore/shell/CMakeLists.txt [hunk 13]
```diff
@@ -41,7 +41,7 @@ if (WIN32)
     add_dependencies(jsc jscLib)
 endif ()
 
-if (NOT WIN32)
+if (NOT WIN32 AND ENABLE_API_TESTS)
```
Comment (annulen) [inline Source/JavaScriptCore/shell/CMakeLists.txt]: Does it really cause problems?

File: Source/JavaScriptCore/PlatformQt.cmake (hunk 10 above)
Comment (matteoceraico) [inline Source/JavaScriptCore/PlatformQt.cmake]: You're right, just tested and all work as expected

File: Source/WebKit/PlatformQt.cmake (hunk 11 above)
Comment (matteoceraico) [inline Source/WebKit/PlatformQt.cmake]: You're right, just tested and all work as expected

File: Source/JavaScriptCore/shell/CMakeLists.txt (hunk 13 above)
Comment (matteoceraico) [inline Source/JavaScriptCore/shell/CMakeLists.txt]: No, now it doesn't cause problems

**Comment by annulen** [comment]
Hint: run `Tools/Scripts/check-webkit-style -g qtwebkit-stable..` in your local repository to check style - it's the same command that is run by Travis CI
